File: gui/frontend/src/lib/autoImplementFilter.ts
// AutoImplementFilter is the persisted/CLI-shaped representation of the
// auto-implement query. It uses arrays for every multi-select field so
// it maps cleanly onto `tb ls`'s comma-separated flags (TB-289). The
// frontend BoardFilter store keeps `parentEpic: string` because the
// board UI is single-epic-focus today; saving converts it into a
// one-element parents array via boardFilterToAutoImplement.
//
// This file is one-way: BoardFilter → AutoImplementFilter (save path).
// There's no inverse helper today — restoring a saved filter into the
// FilterBar isn't part of TB-288's scope and the next reader shouldn't
// add the inverse without a deliberate UX call.

import type { BoardFilter } from '$lib/stores/filter';

export interface AutoImplementFilter {
  search: string;
  types: string[];
  priorities: string[];
  modules: string[];
  sizes: string[];
  tags: string[];
  agents: string[];
  parents: string[];
}

export const emptyAutoImplementFilter: AutoImplementFilter = {
  search: '',
  types: [],
  priorities: [],
  modules: [],
  sizes: [],
  tags: [],
  agents: [],
  parents: [],
};

// boardFilterToAutoImplement serializes the FilterBar state into the
// persisted shape. `parentEpic: ''` becomes `parents: []`; anything
// else becomes a one-element parents array.
export function boardFilterToAutoImplement(f: BoardFilter): AutoImplementFilter {
  return {
    search: f.search.trim(),
    types: [...f.types],
    priorities: [...f.priorities],
    modules: [...f.modules],
    sizes: [...f.sizes],
    tags: [...f.tags],
    agents: [...f.agents],
    parents: f.parentEpic ? [f.parentEpic] : [],
  };
}

// isAutoImplementFilterEmpty mirrors the Go AutoImplementFilter.IsEmpty
// check so the FilterBar Save button, the SettingsPanel needs-filter
// warning, and the +page.svelte missing-prereqs derivation all agree
// on what "no filter saved" means.
export function isAutoImplementFilterEmpty(f: AutoImplementFilter): boolean {
  if (f.search.trim() !== '') return false;
  return (
    f.types.length === 0 &&
    f.priorities.length === 0 &&
    f.modules.length === 0 &&
    f.sizes.length === 0 &&
    f.tags.length === 0 &&
    f.agents.length === 0 &&
    f.parents.length === 0
  );
}

// isBoardFilterActive reports whether the FilterBar's current state
// holds at least one constraint — drives the Save-button enable state.
export function isBoardFilterActive(f: BoardFilter): boolean {
  return !isAutoImplementFilterEmpty(boardFilterToAutoImplement(f));
}

// autoImplementFilterEquals is the equality check that powers the
// "saved" affordance on the FilterBar Save button. Compares structural
// equality of every field including ordered array contents. Ordering
// must match the Go AutoImplementFilter.normalize() output (which only
// trims and drops empty segments, never sorts) — do not sort the
// arrays here, or the "Saved" affordance will flicker after every
// save when the in-memory FilterBar order diverges from disk.
export function autoImplementFilterEquals(a: AutoImplementFilter, b: AutoImplementFilter): boolean {
  if (a.search.trim() !== b.search.trim()) return false;
  return (
    arraysEqual(a.types, b.types) &&
    arraysEqual(a.priorities, b.priorities) &&
    arraysEqual(a.modules, b.modules) &&
    arraysEqual(a.sizes, b.sizes) &&
    arraysEqual(a.tags, b.tags) &&
    arraysEqual(a.agents, b.agents) &&
    arraysEqual(a.parents, b.parents)
  );
}

function arraysEqual(a: string[], b: string[]): boolean {
  if (a.length !== b.length) return false;
  for (let i = 0; i < a.length; i++) if (a[i] !== b[i]) return false;
  return true;
}

// summarize renders the persisted filter as a one-line human-readable
// summary for the SettingsPanel — "Type: bug, improvement · Module:
// gui · Tags: macos". Returns '' for an empty filter so callers can
// switch to a "no filter saved" placeholder.
export function summarizeAutoImplementFilter(f: AutoImplementFilter): string {
  const parts: string[] = [];
  if (f.search.trim() !== '') parts.push(`Search: ${f.search.trim()}`);
  if (f.types.length) parts.push(`Type: ${f.types.join(', ')}`);
  if (f.priorities.length) parts.push(`Priority: ${f.priorities.join(', ')}`);
  if (f.modules.length) parts.push(`Module: ${f.modules.join(', ')}`);
  if (f.sizes.length) parts.push(`Size: ${f.sizes.join(', ')}`);
  if (f.tags.length) parts.push(`Tags: ${f.tags.join(', ')}`);
  if (f.agents.length) parts.push(`Agent: ${f.agents.join(', ')}`);
  if (f.parents.length) parts.push(`Epic: ${f.parents.join(', ')}`);
  return parts.join(' · ');
}

File: gui/app/auto_implement_filter.go
package app

import (
	"encoding/json"
	"strings"
)

// AutoImplementFilter is the persisted shape of the auto-implement query.
// It mirrors the field set TB-289 exposes on `tb ls` so the coordinator
// can hand the filter straight to the CLI without re-implementing
// matching in-process. Field semantics are the same as `tb ls`:
//   - Search: substring match against id+title, case-insensitive.
//   - Types/Priorities/Sizes/Modules/Tags/Agents/Parents: OR-within-field,
//     AND-across-field. Each value matches `tb ls`'s per-field semantics
//     (exact equality for type/priority/size, substring for module, tag
//     name match, normalized parent id, agent enum with "none" sentinel).
//
// The frontend BoardFilter shape uses a single `parentEpic` because the
// board UI is single-epic-focus today; the adapter on save inflates it
// into a one-element Parents slice (or empty). No restore path exists
// yet — the persisted filter is one-way state for the coordinator.
type AutoImplementFilter struct {
	Search     string   `json:"search"`
	Types      []string `json:"types"`
	Priorities []string `json:"priorities"`
	Modules    []string `json:"modules"`
	Sizes      []string `json:"sizes"`
	Tags       []string `json:"tags"`
	Agents     []string `json:"agents"`
	Parents    []string `json:"parents"`
}

// UnmarshalJSON tolerates the legacy text-DSL string form by silently
// resetting to the zero value. loadPreferences logs a one-line migration
// warning when it spots the legacy shape; here we only need to keep the
// decode from failing so a stale `preferences.json` doesn't error.
func (f *AutoImplementFilter) UnmarshalJSON(data []byte) error {
	if len(data) > 0 && data[0] == '"' {
		// Legacy text form: reset to empty filter without erroring.
		*f = AutoImplementFilter{}
		return nil
	}
	type alias AutoImplementFilter
	var a alias
	if err := json.Unmarshal(data, &a); err != nil {
		return err
	}
	*f = AutoImplementFilter(a)
	return nil
}

// IsEmpty reports whether the filter constrains nothing — used to gate
// SetAutoImplementEnabled and to surface the "needs filter" signal in
// AutoImplementStatus.
func (f AutoImplementFilter) IsEmpty() bool {
	if strings.TrimSpace(f.Search) != "" {
		return false
	}
	return len(f.Types) == 0 &&
		len(f.Priorities) == 0 &&
		len(f.Modules) == 0 &&
		len(f.Sizes) == 0 &&
		len(f.Tags) == 0 &&
		len(f.Agents) == 0 &&
		len(f.Parents) == 0
}

// normalize trims whitespace from every value and drops empty segments,
// matching how `tb ls` parses comma-separated input. Returns a copy.
func (f AutoImplementFilter) normalize() AutoImplementFilter {
	return AutoImplementFilter{
		Search:     strings.TrimSpace(f.Search),
		Types:      cleanStringSlice(f.Types),
		Priorities: cleanStringSlice(f.Priorities),
		Modules:    cleanStringSlice(f.Modules),
		Sizes:      cleanStringSlice(f.Sizes),
		Tags:       cleanStringSlice(f.Tags),
		Agents:     cleanStringSlice(f.Agents),
		Parents:    cleanStringSlice(f.Parents),
	}
}

func cleanStringSlice(in []string) []string {
	if len(in) == 0 {
		return nil
	}
	out := make([]string, 0, len(in))
	for _, v := range in {
		v = strings.TrimSpace(v)
		if v == "" {
			continue
		}
		out = append(out, v)
	}
	if len(out) == 0 {
		return nil
	}
	return out
}

// toLsArgs converts the filter into the comma-separated flag arguments
// `tb ls` understands. The status flag is supplied by the caller.
func (f AutoImplementFilter) toLsArgs() []string {
	var args []string
	if joined := strings.Join(f.Types, ","); joined != "" {
		args = append(args, "-T", joined)
	}
	if joined := strings.Join(f.Priorities, ","); joined != "" {
		args = append(args, "-p", joined)
	}
	if joined := strings.Join(f.Modules, ","); joined != "" {
		args = append(args, "-m", joined)
	}
	if joined := strings.Join(f.Sizes, ","); joined != "" {
		args = append(args, "-s", joined)
	}
	if joined := strings.Join(f.Tags, ","); joined != "" {
		args = append(args, "-t", joined)
	}
	if joined := strings.Join(f.Agents, ","); joined != "" {
		args = append(args, "--agent", joined)
	}
	if joined := strings.Join(f.Parents, ","); joined != "" {
		args = append(args, "--parent", joined)
	}
	if s := strings.TrimSpace(f.Search); s != "" {
		args = append(args, "--search", s)
	}
	return args
}
